quickwit/quickwit-query/src/query_ast/user_text_query.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+use std::ops::Bound;
+
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use tantivy::query_grammar::{
+    Occur, UserInputAst, UserInputBound, UserInputLeaf, UserInputLiteral,
+};
+use tantivy::schema::Schema;
+
+use crate::not_nan_f32::NotNaNf32;
+use crate::query_ast::tantivy_query_ast::TantivyQueryAst;
+use crate::query_ast::{self, BuildTantivyAst, QueryAst};
+use crate::{DefaultOperator, InvalidQuery, JsonLiteral};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct UserTextQuery {
+    pub user_text: String,
+    // Set of search fields to search into for text not specifically
+    // targetting a field.
+    //
+    // If None, the default search fields, as defined in the DocMapper
+    // will be used.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub default_fields: Option<Vec<String>>,
+    pub default_operator: DefaultOperator,
+}
+
+impl UserTextQuery {
+    /// Parse the user query to generate a structured QueryAST, without any UserTextQuery node.
+    ///
+    /// The `UserTextQuery` have an optional search_fields property that takes precedence over
+    /// the `default_search_fields`.
+    ///
+    /// In quickwit, the search fields in the `UserTextQuery` are usually supplied with the user
+    /// request.
+    /// The default_search_fields argument on the other hand, is the default search fields defined
+    /// in the `DocMapper`.
+    pub fn parse_user_query(&self, default_search_fields: &[String]) -> anyhow::Result<QueryAst> {
+        let search_fields = self
+            .default_fields
+            .as_ref()
+            .map(|search_fields| &search_fields[..])
+            .unwrap_or(default_search_fields);
+        let user_input_ast = tantivy::query_grammar::parse_query(&self.user_text)
+            .map_err(|_| anyhow::anyhow!("Failed to parse query: `{}`.", &self.user_text))?;
```

Comment:
.with_context

Reply:
It is not possible as the error is not StdErr.